### PR TITLE
Update to hypothesis version 5

### PR DESCRIPTION
**Pull request by vxgmichel**

### CODE REVIEW

**Comment by Zac-HD** [comment]
:wave: with my maintainer-on-both-projects hat on, how about we fix the `multiple()` issue upstream and then depend on `Hypothesis~=5.x`?  It's a quick fix and Hypothesis releases every PR immediately :smile:

**Review by touilleMan — APPROVED** [review]

File: setup.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
     author_email="emmanuel.leblond@gmail.com",
     license="MPL 2.0",
     packages=find_packages(),
-    install_requires=["trio>=0.11", "hypothesis~=4.0"],
+    install_requires=["trio>=0.11", "hypothesis~=5.0"],
```
Comment (Zac-HD) [inline setup.py]: [`hypothesis>=5.1.2,<6`](https://hypothesis.readthedocs.io/en/latest/changes.html#v5-1-2), surely?

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[tool:pytest]
+filterwarnings =
+     error
+     # Ignore trio deprecation warnings in trio_asyncio
+     ignore:trio|Trio:trio.TrioDeprecationWarning:trio_asyncio\.(.*)
```
Comment (Zac-HD) [inline setup.cfg:5]: The first match is applied, so only `error` will ever have an effect here.

File: setup.py (hunk 1 above)
Comment (vxgmichel) [reply setup.py]: I don't know, the `hypothesis-trio` package is technically compatible with `hypothesis==5.1.1`, only the tests will fail. I don't mind changing it though, let me know what you prefer :)

File: setup.cfg (hunk 2 above)
Comment (vxgmichel) [reply setup.cfg:5]: I can confirm it does work as:
- treat warnings as error
- ... but ignore `TrioDeprecationWarning` from `trio_asyncio`

File: setup.py (hunk 1 above)
Comment (Zac-HD) [reply setup.py]: Personally I prefer to use a version bound which I know works for everything, but it's not a strong view in this case.

**Comment by vxgmichel** [comment]
@Zac-HD I've been investigating the issue with the CI and it is related to weird problem:

https://github.com/HypothesisWorks/hypothesis/blob/11c7c08d1a1b5208890f4bb7e1bf9359fcff80d1/hypothesis-python/src/hypothesis/internal/escalation.py#L34-L52

This implementation of `belongs_to` will consider any module inside of `hypothesis_trio` as part of `hypothesis`, which I think is mostly fine. The problem is that the `hypothesis_trio` tests are in `hypothesis_trio/_tests` so the exception raised in the tests (such as legitimate assertion error) are treated as internal errors. Note that this problem only occur when running the tests inside the installed `hypothesis_trio` package, e.g. `mkdir empty; cd empty; pytest --pyargs hypothesis_trio`. This is exactly what travis ci script does.

Do you have any idea about the proper way to fix this issue?

**Review by Zac-HD — APPROVED** [review]
LGTM :+1:

**Comment by vxgmichel** [comment]
@Zac-HD Any idea about [this](https://github.com/python-trio/hypothesis-trio/pull/5#issuecomment-572549558)? Sorry for the ping but I don't really know how to deal with this issue.

**Comment by Zac-HD** [comment]
Err, `belongs_to` *shouldn't* be treating hypothesis_trio as part of Hypothesis... I'll take a look.

**Comment by vxgmichel** [comment]
> Err, belongs_to shouldn't be treating hypothesis_trio as part of Hypothesis... I'll take a look.

I'm not sure about that :thinking:  `hypothesis_trio.stateful` is mostly a copy-paste of `hypothesis.stateful` so any exception raised in there should be considered as internal right?

**Comment by Zac-HD** [comment]
On the basis that I don't want to see those problems on the Hypothesis issue tracker, no, it shouldn't!

**Comment by vxgmichel** [comment]
> On the basis that I don't want to see those problems on the Hypothesis issue tracker, no, it shouldn't!

Alright, then it should easily be fixed with something like:
```diff
diff --git a/hypothesis-python/src/hypothesis/internal/escalation.py b/hypothesis-python/src/hypothesis/internal/escalation.py
index 26ff4bda3..2ff36eaa8 100644
--- a/hypothesis-python/src/hypothesis/internal/escalation.py
+++ b/hypothesis-python/src/hypothesis/internal/escalation.py
@@ -44,7 +44,8 @@ def belongs_to(package):
             return cache[ftype][filepath]
         except KeyError:
             pass
-        result = os.path.abspath(filepath).startswith(root)
+        abspath = os.path.abspath(filepath)
+        result = os.path.commonpath([abspath, root]) == root
         cache[ftype][filepath] = result
         return result
```

**Comment by Zac-HD** [comment]
More directly relevant, I can't reproduce the problem - on my machine adding an `assert False` to the tests gives me a failure, not an internal error.  Can you share more detail on how to reproduce?

**Comment by vxgmichel** [comment]
> Even simpler would be to ensure that root is an abspath 

I don't think that is problem, see this pdb trace: 
```
> /home/vinmic/miniconda/envs/fresh/lib/python3.7/site-packages/hypothesis/internal/escalation.py(48)accept()
-> result = os.path.abspath(filepath).startswith(root)
(Pdb) filepath
'/home/vinmic/miniconda/envs/fresh/lib/python3.7/site-packages/hypothesis_trio/_tests/test_async_stateful.py'
(Pdb) root
'/home/vinmic/miniconda/envs/fresh/lib/python3.7/site-packages/hypothesis'
```

> want to open another upstream PR?

Alright I'll do that.

> More directly relevant, I can't reproduce the problem - on my machine adding an assert False to the tests gives me a failure, not an internal error. Can you share more detail on how to reproduce?

Oh yes sorry about that. The easiest way to reproduce the issue is to follow the `ci/travis.sh` script. For instance:
```shell
cd hypothesis-trio
pip install .
mkdir empty
cd empty
pytest --pyargs hypothesis_trio -k test_trio_style_failing_state_machine
```
The tests will fail because the captured stdout does not contain the right output. I had to dig deep into hypothesis code to realize that this difference came from `belongs_to`.

A simpler way to reproduce the underlying issue is simply to test `is_hypothesis_file`:
```python
In [1]: from hypothesis.internal.escalation import is_hypothesis_file                         
In [2]: from hypothesis_trio import stateful                                                  
In [3]: is_hypothesis_file(stateful.__file__)                                                 
Out[3]: True
```

**Comment by Zac-HD** [comment]
Ah, now I see: my local editable installs were masking the problem, but if your package name starts with `hypothesis` and both are installed in `site-packages` then the hypothesis path string is a prefix of the other.

Proper solution is indeed an upstream fix, which should treat paths as paths not strings :man_facepalming:

**Comment by vxgmichel** [comment]
@Zac-HD I just pushed a PR: https://github.com/HypothesisWorks/hypothesis/pull/2313
I add to use `pathlib.Path.relative_to` since `os.path.commonpath` is not available on all plateforms.

**Comment by vxgmichel** [comment]
I fixed a problem in the tests with python 3.5 but there seems to be another issue:

> The job running on agent Azure Pipelines 14 ran longer than the maximum time of 60 minutes.

**Comment by vxgmichel** [comment]
@Zac-HD Merged! Thanks for your help :)

**Pull request merged**